---
title: "segAnalysis (dm3)"
author: "MGunady"
date: "February 16, 2017"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("ggplot2")
library(stringr)

# Multiple plot function
#
# ggplot objects can be passed in ..., or to plotlist (as a list of ggplot objects)
# - cols:   Number of columns in layout
# - layout: A matrix specifying the layout. If present, 'cols' is ignored.
#
# If the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE),
# then plot 1 will go in the upper left, 2 will go in the upper right, and
# 3 will go all the way across the bottom.
#
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

```

## TraSQ Analysis

This analysis is done using dm3 transcriptome. L values used are {40, 100, 108, 250, 1000, 10000}

```{r settings}
ks = c(40, 100, 108, 250, 1000, 10000)
experiment = "dm3_segs"
```

******

## Transcriptome Characteristics

Some simple plots to understand the transcriptome data. Just for reference if needed in further analysis.

Alternative Splicing Events are defined and counted as follows:
Given the ordered set of (disjoint) exons, if an exon is spliced into a tx which is missed in the following exon, that means a branch in the Splice Graph, hence a splice event.


```{r tx_analysis, echo=FALSE, cache=TRUE}
exper1 = "transcriptome"
gene_txs_count = read.table(paste(exper1, ".fa.gene_segs_count", sep=""))

num_txs = gene_txs_count$V2
ggplot(data.frame(num_txs=num_txs), aes(x = num_txs)) + geom_histogram() + scale_y_log10() +
  ggtitle("Histogram of Number of txs per gene, counts are log scale") +
  theme(plot.title = element_text(hjust = 0.5))


exper = paste(experiment, 100, sep="_")
gene_splices_segs_count = read.table(paste(exper, ".fa.gene_splices_segs_count", sep=""))
num_splices = gene_splices_segs_count$V2
ggplot(data.frame(num_splices=num_splices), aes(x = num_splices)) + geom_histogram() + scale_y_log10() +
  ggtitle("Histogram of Number of Splicing Events per gene, counts are log scale") +
  theme(plot.title = element_text(hjust = 0.5))


plot(num_txs, num_splices, pch=20, main="Number of Txs vs. Number of Splices")

print(paste("GeneIDs with num_txs > 200:", paste(gene_txs_count$V1[num_txs > 200], collapse = ' , ')))


```

******

## Segs Lengths vs. Txs Lengths

The red dotted line represents the L value. 

The higher L the more closer the segments lengths become to the txs lengths. When L = 10,000, since most transcripts are shorter than L, segments almost span the entire tx, so alignment using L = 10,000 should (almost) give the same results as aligning against the transcriptome. 

```{r len_segs_vs_len_txs, echo=FALSE, cache=TRUE}

################################
## Segs Lengths vs. Txs Lengths
################################
exper1 = "transcriptome"
txs_lens = read.table(paste(exper1, "_lens.txt", sep=""))
txs_ones_lens = read.table(paste(exper1, "_lens_fout.txt", sep=""))

df_list = list()

h = hist(log(txs_lens$V2, base = 2))
par(mfrow=c(2,3))
ks=c(40,108,1000,10000)
for(k in ks) {
  exper = paste("dm3_segs", k, sep="_")
  segs_lens = read.table(paste(exper, "_lens.txt", sep=""))
  segs_ones_lens = read.table(paste(exper, "_lens_fout.txt", sep=""))
  
  df_list[[k]] = rbind(data.frame(lens = txs_lens$V2, l=k, label=paste("L=",k, sep=""), type="Transcripts"),
                       #data.frame(lens = txs_ones_lens$V1, l=k, label=paste("L=",k, sep=""), type="F-Txs"),
                       data.frame(lens = segs_lens$V2, l=k, label=paste("L=",k, sep=""), type="Segments")
                       #,data.frame(lens = txs_ones_lens$V1, l=k, label=paste("L=",k, sep=""), type="F-Segs")
                       )
  #Histogram Grey Color
  h = hist(log(segs_lens$V2, base = 2), col=rgb(0.1,0.1,0.1,0.5), main=paste("L=", k, sep=""),
       xlab="Length of sequences (log scale)", ylab="Frequency", breaks=h$breaks)
  hist(log(txs_lens$V2, base = 2), col=rgb(0.8,0.8,0.8,0.5), add=T, breaks = h$breaks)
  hist(log(txs_ones_lens$V1, base = 2), col=rgb(0.8,0.2,0.2,0.5), add=T, breaks = h$breaks)
  hist(log(segs_ones_lens$V1, base = 2), col=rgb(0.2,0.2,0.8,0.5), add=T, breaks = h$breaks)
  abline(v = log(k, base=2), col="red", lwd=1, lty=2)
  #if(k == ks[1]) {
    #legend('topright', col=c(rgb(0.1,0.1,0.1,0.5), rgb(0.8,0.8,0.8,0.5), "red"),
    #     c("Segments", "Transcripts", "L"), lty=1, cex = 0.5, lwd=3)
  #}
  box()
}
df = do.call(rbind, df_list)

ggplot(df, aes(x=log2(lens), colour=type, fill=type)) +
  geom_histogram(bins=20, position="dodge") +
  geom_vline(aes(xintercept = log2(l)), linetype = "dashed", colour = "blue") +
  #geom_density()+
  #scale_y_log10() +
  facet_wrap(~label) +
  ggtitle("(dm3 genome)") +
  labs(y = "Frequency", x="Sequence Length (log scale)") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        legend.position="top", legend.title = element_blank())

ggsave("dm3_segs_lens.pdf", width = 6, height = 4)

```

******

## Per Gene: Number of Segs vs. Number of Txs

Since the scatter plot is dense, different plots are used in here!

The dotted line is a diagonal line; num_segs = num_txs

```{r num_segs_vs_num_txs, echo=FALSE, cache=TRUE}

##################################################
## Per Gene: Number of Segments vs. Number of Txs
##################################################
par(mfrow=c(2,3))
df_list = list()
exper1 = "transcriptome"
gene_txs_count = read.table(paste(exper1, ".fa.gene_segs_count", sep=""))
for(k in ks) {
  exper = paste(experiment, k, sep="_")
  gene_segs_count = read.table(paste(exper, ".fa.gene_segs_count", sep=""))
  num_txs = gene_txs_count$V2
  num_segs = gene_segs_count$V2
  
  # Filtered
  #num_segs = num_segs[num_txs > 1]
  #num_txs = num_txs[num_txs > 1]
  
  plot(num_txs, num_segs, pch=20, main=paste("L=", k, sep=""))
  abline(a=0,b=1, lwd=1, lty=2)
  
  #hist(num_segs/num_txs, main="num_segs/num_txs, +1-txs genes")
  
  
  df = data.frame(num_segs = num_segs, num_txs = num_txs, L = paste("L=", k,sep=""))
  df_list[[k]] = df
}

# par(mfrow=c(1,1))
# ps = list()
# for(k in ks) {
#   ps[[k]] <- ggplot(df_list[[k]], aes(num_txs, num_segs)) + stat_binhex()
# }
# multiplot(ps[[40]], ps[[250]], ps[[100]], ps[[1000]], ps[[108]], ps[[10000]], cols=3)

df = do.call(rbind, df_list)
ggplot(df, aes(x=num_txs, y=num_segs)) + geom_point(shape=20, size=2) + facet_wrap(~L) + 
  geom_smooth(method = "lm", se = FALSE) +
  geom_abline(slope = 1, linetype = "dashed", colour = "black") +
  ggtitle("(dm3 genome)") +
  labs(y = "Number of Segments", x="Number of Transcripts)")  +
  theme_minimal()+
  theme(plot.title = element_text(hjust = 0.5),
        legend.position="top")

ggsave("dm3_segs_num.pdf", width = 6, height = 4)

par(mfrow=c(1,1))
ggplot(df, aes(num_txs, num_segs)) + stat_binhex() +facet_wrap(~L, scales = "free")

```

The scatter plots I think show expected results. 

It may be worth noting that from the plot L = 10,000, it seems transcripts generating multiple segments (note these txs are also longer than L) belong to genes with num_txs of summary:

`summary(num_txs[num_segs/num_txs > 1])`
```{r, echo=FALSE}
df = df_list[[10000]]
summary(df$num_txs[df$num_segs/df$num_txs > 1])
```

Histogram of the num_segs/num_txs ratio (filtered by genes with more than 1 tx)

```{r num_segs_vs_num_txs_hist, echo=FALSE, cache=TRUE, dependson='num_segs_vs_num_txs'}
par(mfrow=c(2,3))
for(k in ks) {
  exper = paste(experiment, k, sep="_")
  gene_segs_count = read.table(paste(exper, ".fa.gene_segs_count", sep=""))
  num_txs = gene_txs_count$V2
  num_segs = gene_segs_count$V2
  
  # Filtered
  num_segs = num_segs[num_txs > 1]
  num_txs = num_txs[num_txs > 1]
  
  hist(num_segs/num_txs, main="num_segs/num_txs, num_txs > 1")
}

```

******

## Per Gene: Number of Segs vs. Number of Splicing Events

```{r num_segs_vs_num_splices, echo=FALSE, cache=TRUE}

######################################################
## Per Gene: Number of Segments vs. Number of Splices
######################################################
par(mfrow=c(2,3))
df_list = list()
for(k in ks) {
  exper = paste(experiment, k, sep="_")
  gene_splices_segs_count = read.table(paste(exper, ".fa.gene_splices_segs_count", sep=""))
  #gene_splices_segs_count = gene_splices_segs_count[gene_splices_segs_count$V2 < 200 & gene_splices_segs_count$V2 > 2, ]
  num_splices = gene_splices_segs_count$V2
  num_segs = gene_splices_segs_count$V3
  
  
  plot(log(num_splices, base = 2), log(num_segs, base = 2), pch=20, main=exper,  col="#00000033")
  abline(a=0,b=1, lwd=1, lty=2)
  
  # Filtered
  filt = num_splices > 0 & num_segs > 1
  num_segs = num_segs[filt]
  num_splices = num_splices[filt]
  
  df = data.frame(log2_num_segs = log(num_segs, base = 2), log2_num_splices = log(num_splices, base = 2), L = paste("L =", k))
  df_list[[k]] = df
}
df = do.call(rbind, df_list)
par(mfrow=c(1,1))
ggplot(df, aes(log2_num_splices, log2_num_segs)) + stat_binhex() +facet_wrap(~L, scales = "free") +
  ggtitle("Filtered data: num_splices > 0 & num_segs > 1") +
  theme(plot.title = element_text(hjust = 0.5))

par(mfrow=c(2,3))
for(k in ks) {
  exper = paste(experiment, k, sep="_")
  gene_splices_segs_count = read.table(paste(exper, ".fa.gene_splices_segs_count", sep=""))
  num_splices = gene_splices_segs_count$V2
  num_segs = gene_splices_segs_count$V3

  # Filtered
  filt = num_splices > 0 & num_segs > 1
  num_segs = num_segs[filt]
  num_splices = num_splices[filt]

  hist(num_segs/num_splices, main="num_segs/num_splices, filtered")
  
}

```

******

## Per Tx: Number of Segs vs. Tx Length


```{r num_segs_vs_tx_lens, echo=FALSE, cache=TRUE}

############################################
## Per Tx: Number of Segments vs. Tx Length
############################################
par(mfrow=c(2,3))
df_list = list()
for(k in ks) {
  exper = paste(experiment, k, sep="_")
  tx_lens_segs_count = read.table(paste(exper, ".fa.tx_lens_segs_count", sep=""))
  tx_lens = tx_lens_segs_count$V2
  num_segs = tx_lens_segs_count$V3
  
  plot(log(tx_lens, base = 2), num_segs, main=paste("num_segs vs. log_tx_len L=", k, sep=""), pch=20)
  
  # Filtered
  filt = num_segs > 1
  num_segs = num_segs[filt]
  tx_lens = tx_lens[filt]
  
  df = data.frame(num_segs = num_segs, log_tx_len = log(tx_lens, base = 2), L = paste("L =", k))
  df_list[[k]] = df
}
df = do.call(rbind, df_list)
par(mfrow=c(1,1))
ggplot(df, aes(log_tx_len, num_segs)) + stat_binhex() +facet_wrap(~L, scales = "free") +
  ggtitle("Filtered data: num_segs > 1") +
  theme(plot.title = element_text(hjust = 0.5))

par(mfrow=c(2,3))
for(k in ks) {
  exper = paste(experiment, k, sep="_")
  tx_lens_segs_count = read.table(paste(exper, ".fa.tx_lens_segs_count", sep=""))
  tx_lens = tx_lens_segs_count$V2
  num_segs = tx_lens_segs_count$V3

  # Filtered
  filt = num_segs > 1
  num_segs = num_segs[filt]
  tx_lens = tx_lens[filt]

  hist(num_segs/log(tx_lens, base = 2), main="#segs/log_txlen, num_segs>1")
  
}

```


******

## Per x-tx gene: Number of Segs, Number of Splices

For genes with x transcripts, the dist. of the number of segs and splices are shown.
Selected values of x are: `num_txs = c(1, 2, 3, 5, 7, 10, 20, 30, 40, 45, 60, 70)`

```{r x_num_segs_vs_num_txs, echo=FALSE, cache=TRUE}

##################################################
## Per Gene: Number of Segments, Number of Splices
##################################################

boxplot_xtx <- function(k, xs) {
  df_list = list()
  exper1 = "transcriptome"
  gene_txs_count = read.table(paste(exper1, ".fa.gene_segs_count", sep=""))
  exper = paste(experiment, k, sep="_")
  gene_splices_segs_count = read.table(paste(exper, ".fa.gene_splices_segs_count", sep=""))
  
  for(x in xs) {
    x_genes = gene_txs_count$V1[gene_txs_count$V2 == x]
    
    x_num_splices = gene_splices_segs_count$V2[gene_splices_segs_count$V1 %in% x_genes]
    x_num_segs = gene_splices_segs_count$V3[gene_splices_segs_count$V1 %in% x_genes]
    
    df = data.frame(num_segs = x_num_segs, num_splices = x_num_splices, num_txs = x, L = paste("L =", k))
    
    #boxplot(num_segs~num_txs,data=df)
    
    df_list[[x]] = df
  }
  df = do.call(rbind, df_list)
  #boxplot(num_segs~x,data=df)
  return(df)
}

par(mfrow=c(1,1))
xs=c(1, 2, 3, 5, 7, 9, 11, 15, 21, 25, 31, 75)

df_list = list()
for(k in ks) {
  df_list[[k]] = boxplot_xtx(k, xs)
}
df = do.call(rbind, df_list)
ggplot(df, aes(factor(num_txs), num_segs)) + geom_boxplot() + facet_wrap(~L, scales = "free") +
  ggtitle("Number of Segs dist. for x-txs genes") +
  theme(plot.title = element_text(hjust = 0.5))

ggplot(df[df$L == "L = 100", ], aes(factor(num_txs), num_splices)) + geom_boxplot() +
  ggtitle("Number of Splicing Events dist. for x-txs genes") +
  theme(plot.title = element_text(hjust = 0.5))

f_df = df[df$L == "L = 100", ]

f_df$num_txs = paste("num_txs =", str_pad(f_df$num_txs, 2, pad = "0"))
  
ggplot(f_df, aes(factor(num_splices), num_segs)) + geom_boxplot() + facet_wrap(~num_txs, scales = "free") +
  ggtitle("Number of Segs vs. Number of splices for x-txs genes, L = 100") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=7))


```

